### ISSUE TITLE
Question about computational efficiency and numerical stability in Grad-CAM++

### ISSUE THREAD

**Issue by mlerma54** [jacobgil/pytorch-grad-cam#236]
In the code for Grad-CAM++ I see the following line:

        aij = grads_power_2 / (2 * grads_power_2 +
                               sum_activations[:, :, None, None] * grads_power_3 + eps)

which implements the following equation from the orignal Grad-CAM++ paper:

![img 1](https://user-images.githubusercontent.com/5579488/163826983-474d6fdb-858e-4d93-97b0-ef809d7e2dbd.png)

However that formula can be simplified by dividing numerator and denominator by the squared gradient:

![img2](https://user-images.githubusercontent.com/5579488/163826984-02649485-170b-439c-b87b-73fd5a9abeee.png)

which produces a simpler, more efficient, and more numerically stable code:

        aij = 1 / (2  +  sum_activations[:, :, None, None] * grads + eps)

Why spend CPU time unnecessarily computing powers of gradients, which are also vulnerable to over and underflow?

**Comment by jacobgil** [comment]
Good question.
The only reason I did this was to stick to original formula for readability and clarity.
There is a trade of here between simplifying it here, and making potential bugs along the way.

But yeah - looks like we can just use the line you proposed !
Except pixels where the gradient is zero - then the result should be zero in the original formula, but won't be zero in yours.

Did you test it?
Does it looks good?
Did you see examples where it was actually numerically stable in the reformulation and didn't work in the original one ?

**Comment by mlerma54** [comment]
There is another line that takes care of the of zero gradient case:

        aij = np.where(grads != 0, aij, 0)

I have been experimenting in the last year or so with various gradient-based attribution methods and occasionally the gradients become very small ("vanishing gradients" problem addressed by some methods such as Integrate Gradients). I don't remember cases of overflow (that would trigger some sort of exception).

I just started looking at Grad-CAM++ recently. The math in the original paper is hard to follow (e.g. after taking partial derivatives in their equation 7 I do not get their equation 8, but one with an extra summation and partial cross-derivatives). Some of their assumptions such as redefining Y^c = exp(S^c) had me scratching my head (a redefinition such as Y^c = exp(\lambda S^c) for any positive \lambda is just as legitimate but can produce very different results). The original Y^c can be replaced with Y^c + any linear combinations of the A_{Ij}^{k} and (since second order derivatives kill linearites) their method produces the same alphas even though they cannot anymore be solutions to equation (7). I am still trying to understand why the method works even though ts mathematical foundations are not quite clear.

**Comment by mlerma54** [comment]
I appreciate your answers. I have another question. What is the role of 'eps' in the denominator?

        eps = 0.000001
        aij = grads_power_2 / (2 * grads_power_2 +
                               sum_activations[:, :, None, None] * grads_power_3 + eps)

It looks like the kind of small constant used to prevent division by zero, but it works only if the denominator is supposed to be non-negative, which is not the case here. The official code posted by the authors does not use it, the way they prevent division by zero is by directly checking when the denominator becomes zero (third line in this code): 

	alpha_num = conv_second_grad[0]
	alpha_denom = conv_second_grad[0]*2.0 + conv_third_grad[0]*global_sum.reshape((1,1,conv_first_grad[0].shape[2]))
	alpha_denom = np.where(alpha_denom != 0.0, alpha_denom, np.ones(alpha_denom.shape))
	alphas = alpha_num/alpha_denom

Source: https://github.com/adityac94/Grad_CAM_plus_plus/blob/master/misc/utils.py

Interestingly the value of the 'eps' constant seems to have an impact, to the point I wonder whether it should be introduced as an hyper-parameter to be adjusted empirically. For one of the Imagenet sample images I am using for testing the aij obtained at the last layer of a pretrained VGG19 the elements of aij (alphas in the author's code) range from -11.087788 to 26.845564 with eps = 0.000001 (per your code), and from -53.552486  to 81.40964 if eps = 0. The distributions of those values are also very different (as reveled by displaying them in a whisker-plot) - they tend to concentrate more around 0.5 when eps is smaller, and spread out if eps is set to a larger value.

**Comment by jacobgil** [comment]
Hi,
Yes, this is a bit of a deviation. And yes it could be a hyper-parameter, although I'm surprised it has an impact on the final visualization. If it does, it might mean that the method is too sensitive.
I think we could replace with the 0 check directly.

Closing this issue for now, please re open if you have any other question about this.
